lib/nostrum/shard/session.ex

```diff
@@ -81,59 +110,76 @@ defmodule Nostrum.Shard.Session do
 
     case from_handle do
       {new_state, reply} ->
-        {:reply, {:binary, reply}, new_state}
+        :ok = :gun.ws_send(state.conn, {:binary, reply})
+        {:noreply, new_state}
 
       new_state ->
-        {:ok, new_state}
+        {:noreply, new_state}
     end
   end
 
-  def handle_cast({:status_update, payload}, state) do
-    {:reply, {:binary, payload}, state}
-  end
+  def handle_info(
+        {:gun_down, conn, protocol, reason, _killed_streams, _unprocessed_streams},
+        state
+      ) do
+    Logger.warn(fn -> "connection down (#{protocol} #{reason}), attempting reconnect" end)
 
-  def handle_cast({:request_guild_members, payload}, state) do
-    {:reply, {:binary, payload}, state}
-  end
-
-  def handle_cast(:heartbeat, %{heartbeat_ack: false} = state) do
-    Logger.warn("heartbeat_ack not received in time, disconnecting")
-    {:close, state}
-  end
+    {:ok, :cancel} = :timer.cancel(state.heartbeat_ref)
+    :ok = :gun.close(conn)
 
-  def handle_cast(:heartbeat, state) do
-    {:ok, ref} =
-      :timer.apply_after(state.heartbeat_interval, WebSockex, :cast, [state.conn_pid, :heartbeat])
+    worker = connect(state.gateway)
```

Comment:
I didn't see this handler before the prior review comments, but I'll leave them there regardless for posterity. Like I said elsewhere, we shouldn't try to reconnect the ws in this callback. Let it restart itself and handle it properly.